recipes/generate.py

```diff
@@ -0,0 +1,176 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+import sys
+import time
+from typing import Any, Dict
+
+import torch
+from omegaconf import DictConfig
+
+from torch import nn
+
+from torchtune import config, utils
+
+logger = utils.get_logger("DEBUG")
+
+
+class InferenceRecipe:
+    def __init__(self, cfg: DictConfig) -> None:
+        self._device = utils.get_device(device=cfg.device)
+        self._dtype = utils.get_dtype(dtype=cfg.dtype)
+
+        utils.set_seed(seed=cfg.seed)
+
+    def load_checkpoint(self, checkpointer_cfg: DictConfig) -> Dict[str, Any]:
+        checkpointer = config.instantiate(checkpointer_cfg)
+        checkpoint_dict = checkpointer.load_checkpoint()
+        return checkpoint_dict
+
+    def setup(self, cfg: DictConfig) -> None:
+        ckpt_dict = self.load_checkpoint(cfg.checkpointer)
+        self._model = self._setup_model(
+            model_cfg=cfg.model,
+            model_state_dict=ckpt_dict[utils.MODEL_KEY],
+        )
+        self._tokenizer = config.instantiate(cfg.tokenizer)
+
+    def _setup_model(
+        self,
+        model_cfg: DictConfig,
+        model_state_dict: Dict[str, Any],
+    ) -> nn.Module:
+        with utils.set_default_dtype(self._dtype), self._device:
+            model = config.instantiate(model_cfg)
+
+        model.load_state_dict(model_state_dict)
+
+        # Validate model was loaded in with the expected dtype.
+        utils.validate_expected_param_dtype(model.named_parameters(), dtype=self._dtype)
+        logger.info(f"Model is initialized with precision {self._dtype}.")
+
+        # Ensure the cache is setup on the right device
+        with self._device:
+            model.setup_caches(max_batch_size=1, dtype=self._dtype)
+        return model
+
+    def _multinomial_sample_one_no_sync(self, probs_sort):
+        q = torch.empty_like(probs_sort).exponential_(1)
+        return torch.argmax(probs_sort / q, dim=-1, keepdim=True).to(dtype=torch.int)
+
+    def _logits_to_probs(self, logits: torch.Tensor, temperature: float, top_k: int):
```

Comment:
We're pretty inconsistent with where our generation code lives. I'd be in favor of putting it all in one file.

Reply:
I don't want to generalize this code prematurely. I think for now I expect this to be limited to this recipe, but we can generalize as a follow up if that makes sense

Reply:
A couple questions here: 

1) Does this obviate the need for `logits_transforms.py`? Seems we are now handling all of that in the recipe. (As a follow-up, if we are gonna keep it, can we at least drop the `LogitsTransform` ABC? It's literally just a `Callable[FloatTensor, FloatTensor]`)
2) I notice we are missing top p sampling, which we otherwise have support for. Any particular reason for omitting it?

Reply:
I removed the logits_transform.py file completely. For the support for top_p, I didn't really find a good reference. Maybe we can add back if its needed?